File: CS_Samples/bmp2imm.cs
///////////////////////////////////////////////////////////////////////////////
///
/// FuryUtils Sample
/// 
/// Converting a windows Bmp format file to an Imm format file.
/// 
/// using:
/// 
///    Bmp
///         class for reading/writing images in Bmp format.
///    FuryException
///         exception class for reporting errors.
///
///
/// Suggested inputs from the testassets: pal8out.bmp
/// 
///////////////////////////////////////////////////////////////////////////////

using carbon14.FuryStudio.Utils;

namespace carbon14.FuryStudio.CS_Samples
{
    internal class bmp2imm
    {
        public static void Execute(string bmpFileName, string immFileName, string pamFileName)
        {
            Console.WriteLine("bmp2imm Sample");

            try
            {
                // Read bmp file into buffer.
                byte[] bmpBuffer;
                using (FileStream fs = new FileStream(bmpFileName, FileMode.Open))
                {
                    bmpBuffer = new byte[fs.Length];
                    fs.Read(bmpBuffer, 0, bmpBuffer.Length);
                }

                byte[]? immBuffer;
                byte[]? vgaBuffer;

                // Create a Bmp object from the buffer.
                // and get buffers for the raw image and palette data.
                using (Bmp bmp = new Bmp(bmpBuffer))
                {
                    immBuffer = bmp.ImmBuffer;
                    vgaBuffer = bmp.VgaBuffer;
                }

                // Write the raw image buffer into a file
                if (immBuffer != null)
                {
                    using (FileStream fs = new FileStream(immFileName, FileMode.Create))
                    {
                        fs.Write(immBuffer, 0, immBuffer.Length);
                    }
                }

                // Write the raw palette buffer into a file
                if (vgaBuffer != null)
                {
                    using (FileStream fs = new FileStream(pamFileName, FileMode.Create))
                    {
                        fs.Write(vgaBuffer, 0, vgaBuffer.Length);
                    }
                }
            }
            catch (FuryException ex)
            {
                Console.WriteLine($"Error: {ex.ErrorCode} {ex.Message}");
            }
        }
    }
}

File: Utils/Cpp_Samples/bmp2imm.cpp
///////////////////////////////////////////////////////////////////////////////
// FuryUtils Sample
//
// Converting a windows Bmp format file to an Imm format file
//
// using:
//
//    Bmp
//       class for reading / writing images in Bmp format.
//    Exceptions::Exception
//       exception class for reporting errors.
//
///////////////////////////////////////////////////////////////////////////////

#include <ios>
#include <fstream>
#include "../include/FuryUtils.hpp"

void bmp2imm(const char * bmpFileName, const char * immFileName, const char * pamFileName) {
	printf("bmp2imm Sample\n");
	
	try
	{
		// Get a stream for the Bmp format file and load it into a buffer
		std::ifstream bmpStream(bmpFileName, std::ios::binary | std::ios::ate);
		std::streamsize size = bmpStream.tellg();
		bmpStream.seekg(0, std::ios::beg);

		std::vector<uint8_t> bmpBuffer((uint32_t)size);
		bmpStream.read((char *)(bmpBuffer.data()), size);

		// Create a Bmp object from the input stream
		FuryUtils::Image::Bmp bmp(bmpBuffer);

		// Get a buffer of the raw image file and save it to a stream
		std::vector<uint8_t> immBuffer;
		bmp.ImmBuffer(immBuffer);

		std::ofstream immStream(immFileName, std::ios::binary | std::ios::trunc);
		immStream.write((char *)(immBuffer.data()), immBuffer.size());

		// Get a buffer of the palette and save it to a stream
		std::vector<uint8_t> pamBuffer;
		bmp.PamBuffer(pamBuffer);

		std::ofstream pamStream(pamFileName, std::ios::binary | std::ios::trunc);
		pamStream.write((char *)(pamBuffer.data()), pamBuffer.size());

		return;
	}
	catch (FuryUtils::Exceptions::Exception e)
	{
		printf("Error:\n\n%d %s\n", e._errorCode, e._errorString.c_str());
		return;
	}
	return;
}

